lib/_http_agent.js

```diff
@@ -202,12 +202,25 @@ Agent.prototype.addRequest = function addRequest(req, options, port/* legacy */,
     this.sockets[name] = [];
   }
 
-  const freeLen = this.freeSockets[name] ? this.freeSockets[name].length : 0;
+  const freeSockets = this.freeSockets[name];
+  let socket;
+  while (!socket && freeSockets && freeSockets.length) {
+    socket = freeSockets.shift();
+
+    // TODO(ronag): Might want to look into using writable, readable and/or
+    // _writableState.errored, _readableState.errored.
+    if (socket.destroyed) {
+      socket = null;
+    }
+    // don't leak
+    if (!freeSockets.length)
+      delete this.freeSockets[name];
+  }
+
+  const freeLen = freeSockets ? freeSockets.length : 0;
```

Comment:
freeLen used to include the length of the socket that was removed, however I think that was a mistake as well and it's only used for debug logging.